### DISCUSSION TITLE
Feature complete (but almost certainly bug infested) 2.3.0 is DANGEROUSLY CLOSE

### DISCUSSION THREAD

**Discussion by SpenceKonde** [SpenceKonde/megaTinyCore#377] [Future Development Discussion]
Full functionality means:
* ADC can be used to do analogRead. 
* ADC on 0/1/2-series can be used to take burst accumulated readings using the analogReadEnh() (Enhanced), and either give you the accumuilated value for your processing needs, or decimate them for you to just get you the more accurate readings you want. 
* On the 2-series, analogReadDiff will allow taking differential ADC readings with the same oversampling facilities. (analogReadEnh and analogReadDiff will be ported to DxCore once the design flaws are found here)
* Configure whether PGA, references, etc are kept on or not (function call to set LOWLAT and PGAEN, and enable auto-turnoff of PGA after using it or not. 
* Series/triggered conversions are not supported - no plans to support them within the core as there 
* Most of the code written in one sitting after studying datasheet for weeks. It's pretty ugly and needs some cleanup....
* Oh, and the comparator library is ported too, there's that newer version of the pyupdi style programmerr.....

**Comment by jvasileff** [comment]
That's great. I'm trying it out now.

One question - how do folks usually install `avr-gcc` on the Mac? Using the Arduino IDE's gcc (via cli or the ide) results in:

```
/Users/jvasileff/Documents/Arduino/hardware/megaTinyCore/megaavr/cores/megatinycore/Arduino.h:95:34: error: 'ADC_MUXPOS_DACREF0_gc' undeclared (first use in this function); did you mean 'ADC_MUXPOS_DAC_gc'?
   #define ADC_DACREF0     ADC_CH(ADC_MUXPOS_DACREF0_gc)
                                  ^
```

I am able to compile using a Makefile using `-I path/to/atpack/include` after downloading and unpacking `Atmel.ATtiny_DFP.1.9.337.atpack`, but this is slightly less convenient than the setup for 0 and 1 series chips which works without the atpack.

**Comment by SpenceKonde** [comment]
Uh, when the core is installable through board manager, you just use that. butthere isnt a release of this yet. 

I suspect you can do the same thing  as windows and take an arduino installation with no exposure to board manager (I don;'t know how to do portable mode on mac, but thats what you want - yowant it to use the stock toolchain), so you can then delete that copy of the toolchain and replace it with the package that I distribute with DxCore 

Of course, this is all complicated by the fact that I screwed up the last dxcore rleease and haven't put up the fixed toolchain files yet... = uou probably want the azduino4a version of the toolchain. but I haven't posted it,. =

**Reply by jvasileff** [reply]
> butthere isnt a release of this yet

Sure - I installed commit 572ae53 from GitHub. I'll keep an eye out for an updated toolchain (which I presume would have updated `iotn1626.h`, `iotn1627.h`, etc. files). For now, my DIY approach with the `atpack` seems to work.

**Reply by SpenceKonde** [reply]
New toolchain package is up (azduiino4a). 
```
        {
          "name": "avr-gcc",
          "version": "7.3.0-atmel3.6.1-azduino4a",
          "systems": [
            {
              "size": "41235915",
              "checksum": "SHA-256:23c026f63fe242173a575ff378e46378976cda3cffcc1a8503e3d9178d5cfae0",
              "host": "i686-mingw32",
              "archiveFileName": "avr-gcc-7.3.0-atmel3.6.1-azduino4-i686-w64-mingw32.tar.bz2",
              "url": "https://spencekondetoolchains.s3.amazonaws.com/avr-gcc-7.3.0-atmel3.6.1-azduino4-i686-w64-mingw32.tar.bz2"
            },
            {
              "size": "37878659",
              "checksum": "SHA-256:23becaaf746e8cf3b23ebcd261bf52d5eebcc094c3500fe8d533f0911b870c60",
              "host": "aarch64-linux-gnu",
              "archiveFileName": "avr-gcc-7.3.0-atmel3.6.1-azduino4a-aarch64-pc-linux-gnu.tar.bz2",
              "url": "https://spencekondetoolchains.s3.amazonaws.com/avr-gcc-7.3.0-atmel3.6.1-azduino4a-aarch64-pc-linux-gnu.tar.bz2"
            },
            {
              "size": "34220996",
              "checksum": "SHA-256:dacef5c90b38e6728e2cd161a54133b008352ff6f6513e67ab71f3117331c758",
              "host": "arm-linux-gnueabihf",
              "archiveFileName": "avr-gcc-7.3.0-atmel3.6.1-azduino4a-arm-linux-gnueabihf.tar.bz2",
              "url": "https://spencekondetoolchains.s3.amazonaws.com/avr-gcc-7.3.0-atmel3.6.1-azduino4a-arm-linux-gnueabihf.tar.bz2"
            },
            {
              "size": "36887243",
              "checksum": "SHA-256:a3f4833034b5526a7375b20fed7031e5fd0ff6956703e7b4bcf48e914fb94d6c",
              "host": "i686-linux-gnu",
              "archiveFileName": "avr-gcc-7.3.0-atmel3.6.1-azduino4a-i686-pc-linux-gnu.tar.bz2",
              "url": "https://spencekondetoolchains.s3.amazonaws.com/avr-gcc-7.3.0-atmel3.6.1-azduino4a-i686-pc-linux-gnu.tar.bz2"
            },
            {
              "size": "37530243",
              "checksum": "SHA-256:61c9ad88f9e75664fc6b96350dd4b3f95946e1c6eeca2158504aecc515f475a3",
              "host": "x86_64-apple-darwin14",
              "archiveFileName": "avr-gcc-7.3.0-atmel3.6.1-azduino4a-x86_64-apple-darwin14.tar.bz2",
              "url": "https://spencekondetoolchains.s3.amazonaws.com/avr-gcc-7.3.0-atmel3.6.1-azduino4a-x86_64-apple-darwin14.tar.bz2"
            },
            {
              "size": "37412841",
              "checksum": "SHA-256:60d879e7052db01383cd89c54e46fa20c81ca180e20a8c0a681ffecec20c24d8",
              "host": "x86_64-linux-gnu",
              "archiveFileName": "avr-gcc-7.3.0-atmel3.6.1-azduino4a-x86_64-pc-linux-gnu.tar.bz2",
              "url": "https://github.com/SpenceKonde/DxCore/releases/download/1.3.4/avr-gcc-7.3.0-atmel3.6.1-azduino4a-x86_64-pc-linux-gnu.tar.bz2"
            }
```
High hopes for megaTinyCore 2.3.0 release friday/saturday.

**Reply by jvasileff** [reply]
Thanks, that works.

In theory it seems like `avr-gcc` could be installed to `~/Documents/Arduino/tools/avr-gcc/7.3.0-atmel3.6.1-azduino4a`, but that didn't work for me.

What did work was unpacking the files to `~/Library/Arduino15/packages/MANUAL_INSTALL/tools/avr-gcc/7.3.0-atmel3.6.1-azduino4a` with `MANUAL_INSTALL` being some arbitrary text.

```
mkdir -p ~/Library/Arduino15/packages/MANUAL_INSTALL/tools/avr-gcc/7.3.0-atmel3.6.1-azduino4a
curl https://spencekondetoolchains.s3.amazonaws.com/avr-gcc-7.3.0-atmel3.6.1-azduino4a-x86_64-apple-darwin14.tar.bz2 \
    | tar xvj \
        --strip-components 1 \
        -C ~/Library/Arduino15/packages/MANUAL_INSTALL/tools/avr-gcc/7.3.0-atmel3.6.1-azduino4a
```

I have the core installed (checked out) in `~/Documents/Arduino/hardware/megaTinyCore/`.

**Reply by jvasileff** [reply]
Are the build scripts for the `azduino4a` tool chain published anywhere?
